File: Test/07_merging_annotation/test_merging_annotation.py
"""
CASSIA Test 07: Merging Annotation
==================================
Tests the merge_annotations and merge_annotations_all functions for grouping
cell type annotations at different detail levels.

Usage:
    python test_merging_annotation.py

Note: This test requires batch annotation results. If none exist, it will
run a batch annotation first.
"""

import sys
import time
from pathlib import Path

# Add shared utilities to path
sys.path.insert(0, str(Path(__file__).parent.parent / "shared" / "python"))

from fixtures import get_full_marker_dataframe, get_marker_file_path
from test_utils import (
    setup_cassia_imports,
    load_config,
    setup_api_keys,
    print_test_header,
    print_test_result,
    print_config_summary
)
from result_manager import (
    create_results_dir,
    save_test_metadata,
    save_test_results,
    create_test_metadata,
    get_latest_results
)

# Setup CASSIA imports
setup_cassia_imports()


def run_merging_annotation_test():
    """Test merging annotation functionality."""
    print_test_header("07 - Merging Annotation")

    # Load configuration
    config = load_config()
    print_config_summary(config)

    # Setup API keys
    setup_api_keys()

    # Get LLM settings
    llm_config = config['llm']
    data_config = config['data']

    # Import CASSIA functions
    from tools_function import runCASSIA_batch
    from merging_annotation import merge_annotations, merge_annotations_all
    import pandas as pd

    # Create results directory
    results_dir = create_results_dir("07_merging_annotation")
    print(f"Results will be saved to: {results_dir}")

    # First, we need batch results for merging
    # Check if we have existing batch results
    batch_results_dir = get_latest_results("02_batch_annotation")
    batch_results_file = None

    if batch_results_dir:
        potential_file = batch_results_dir / "batch_results_full.csv"
        if potential_file.exists():
            batch_results_file = str(potential_file)
            print(f"\nUsing existing batch results: {batch_results_file}")

    # If no existing results, run batch annotation first
    if not batch_results_file:
        print("\nNo existing batch results found. Running batch annotation first...")
        marker_df = get_full_marker_dataframe()
        batch_output = str(results_dir / "batch_for_merge")

        runCASSIA_batch(
            marker=marker_df,
            output_name=batch_output,
            n_genes=data_config.get('n_genes', 30),
            model=llm_config.get('model', 'google/gemini-2.5-flash'),
            temperature=llm_config.get('temperature', 0.3),
            tissue=data_config.get('tissue', 'large intestine'),
            species=data_config.get('species', 'human'),
            max_workers=llm_config.get('max_workers', 3),
            provider=llm_config.get('provider', 'openrouter'),
            validator_involvement=config.get('validator', {}).get('default', 'v1')
        )
        batch_results_file = f"{batch_output}_full.csv"

    # Run merging tests
    start_time = time.time()
    errors = []
    status = "error"
    merge_results = {}

    try:
        # Test 1: Single detail level merge (broad)
        print("\n--- Test 1: Single detail level merge (broad) ---")
        broad_output = str(results_dir / "merge_broad.csv")

        result_broad = merge_annotations(
            csv_path=batch_results_file,
            output_path=broad_output,
            provider=llm_config.get('provider', 'openrouter'),
            model=llm_config.get('model', 'google/gemini-2.5-flash'),
            detail_level="broad",
            batch_size=10
        )

        if result_broad is not None and "Merged_Grouping_1" in result_broad.columns:
            print(f"  Broad merge: SUCCESS")
            print(f"  Output rows: {len(result_broad)}")
            print(f"  Sample groupings:")
            for idx, row in result_broad.head(3).iterrows():
                print(f"    {row['True Cell Type']} -> {row['Merged_Grouping_1']}")
            merge_results['broad'] = {
                'status': 'success',
                'output_file': broad_output,
                'num_rows': len(result_broad)
            }
        else:
            errors.append("Broad merge failed - missing Merged_Grouping_1 column")
            merge_results['broad'] = {'status': 'failed'}

        # Test 2: Single detail level merge (detailed)
        print("\n--- Test 2: Single detail level merge (detailed) ---")
        detailed_output = str(results_dir / "merge_detailed.csv")

        result_detailed = merge_annotations(
            csv_path=batch_results_file,
            output_path=detailed_output,
            provider=llm_config.get('provider', 'openrouter'),
            model=llm_config.get('model', 'google/gemini-2.5-flash'),
            detail_level="detailed",
            batch_size=10
        )

        if result_detailed is not None and "Merged_Grouping_2" in result_detailed.columns:
            print(f"  Detailed merge: SUCCESS")
            print(f"  Output rows: {len(result_detailed)}")
            print(f"  Sample groupings:")
            for idx, row in result_detailed.head(3).iterrows():
                print(f"    {row['True Cell Type']} -> {row['Merged_Grouping_2']}")
            merge_results['detailed'] = {
                'status': 'success',
                'output_file': detailed_output,
                'num_rows': len(result_detailed)
            }
        else:
            errors.append("Detailed merge failed - missing Merged_Grouping_2 column")
            merge_results['detailed'] = {'status': 'failed'}

        # Test 3: All detail levels merge (parallel)
        print("\n--- Test 3: All detail levels merge (parallel) ---")
        all_output = str(results_dir / "merge_all.csv")

        result_all = merge_annotations_all(
            csv_path=batch_results_file,
            output_path=all_output,
            provider=llm_config.get('provider', 'openrouter'),
            model=llm_config.get('model', 'google/gemini-2.5-flash'),
            batch_size=10
        )

        expected_columns = ["Merged_Grouping_1", "Merged_Grouping_2", "Merged_Grouping_3"]
        if result_all is not None and all(col in result_all.columns for col in expected_columns):
            print(f"  All levels merge: SUCCESS")
            print(f"  Output rows: {len(result_all)}")
            print(f"  Columns: {', '.join(expected_columns)}")
            print(f"\n  Sample comparison:")
            for idx, row in result_all.head(3).iterrows():
                print(f"    {row['True Cell Type']}:")
                print(f"      Broad:        {row['Merged_Grouping_1']}")
                print(f"      Detailed:     {row['Merged_Grouping_2']}")
                print(f"      Very Detailed: {row['Merged_Grouping_3']}")
            merge_results['all'] = {
                'status': 'success',
                'output_file': all_output,
                'num_rows': len(result_all),
                'columns': expected_columns
            }
        else:
            missing = [col for col in expected_columns if col not in (result_all.columns if result_all is not None else [])]
            errors.append(f"All levels merge failed - missing columns: {missing}")
            merge_results['all'] = {'status': 'failed'}

        # Determine overall status
        if all(r.get('status') == 'success' for r in merge_results.values()):
            status = "passed"
        elif any(r.get('status') == 'success' for r in merge_results.values()):
            status = "partial"
        else:
            status = "failed"

    except Exception as e:
        errors.append(str(e))
        status = "error"
        print(f"\nError: {e}")
        import traceback
        traceback.print_exc()

    duration = time.time() - start_time

    # Save metadata and results
    metadata = create_test_metadata(
        test_name="merging_annotation",
        config=config,
        duration_seconds=duration,
        status=status,
        clusters_tested=["all"],
        errors=errors
    )
    save_test_metadata(results_dir, metadata)

    save_test_results(results_dir, {
        "batch_results_file": batch_results_file,
        "merge_results": merge_results,
        "detail_levels_tested": ["broad", "detailed", "very_detailed", "all"]
    })

    # Print final result
    success = status == "passed"
    print_test_result(success, f"Duration: {duration:.2f}s")

    return success


if __name__ == "__main__":
    success = run_merging_annotation_test()
    sys.exit(0 if success else 1)

File: Test/07_merging_annotation/test_merging_annotation.R
# CASSIA Test 07: Merging Annotation (R)
# ======================================
# Tests the merge_annotations and merge_annotations_all functions via R package.
#
# Usage:
#     Rscript test_merging_annotation.R
#
# Note: This test requires batch annotation results. If none exist, it will
# run a batch annotation first.

# Get script directory (works with Rscript and source())
get_script_dir <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("--file=", args, value = TRUE)
  if (length(file_arg) > 0) {
    return(dirname(normalizePath(sub("--file=", "", file_arg), winslash = "/")))
  }
  for (i in sys.nframe():1) {
    if (!is.null(sys.frame(i)$ofile)) {
      return(dirname(normalizePath(sys.frame(i)$ofile, winslash = "/")))
    }
  }
  return(normalizePath(getwd(), winslash = "/"))
}
script_dir <- get_script_dir()

# Source shared utilities
source(file.path(script_dir, "..", "shared", "r", "test_utils.R"))
source(file.path(script_dir, "..", "shared", "r", "fixtures.R"))
source(file.path(script_dir, "..", "shared", "r", "result_manager.R"))

run_merging_annotation_test <- function() {
  print_test_header("07 - Merging Annotation (R)")

  # Load configuration
  config <- load_config()
  print_config_summary(config)

  # Setup API keys
  setup_api_keys()

  # Setup CASSIA R package
  tryCatch({
    setup_cassia()
    message("CASSIA R package loaded successfully")
  }, error = function(e) {
    message("Error loading CASSIA: ", e$message)
    return(FALSE)
  })

  # Get settings
  llm_config <- config$llm
  data_config <- config$data

  # Create results directory
  results_dir <- create_results_dir("07_merging_annotation")
  cat("Results will be saved to:", results_dir, "\n")

  # Check for existing batch results from Test 02
  batch_results_dir <- get_latest_results("02_batch_annotation")
  batch_results_file <- NULL

  if (!is.null(batch_results_dir)) {
    potential_file <- file.path(batch_results_dir, "batch_results_full.csv")
    if (file.exists(potential_file)) {
      batch_results_file <- potential_file
      cat("\nUsing existing batch results:", batch_results_file, "\n")
    }
  }

  # If no existing results, run batch annotation first
  if (is.null(batch_results_file)) {
    cat("\nNo existing batch results found. Running batch annotation first...\n")
    marker_df <- get_full_marker_dataframe()
    batch_output <- file.path(results_dir, "batch_for_merge")

    CASSIA::runCASSIA_batch(
      marker = marker_df,
      output_name = batch_output,
      n_genes = data_config$n_genes %||% 30,
      model = llm_config$model %||% "google/gemini-2.5-flash",
      temperature = llm_config$temperature %||% 0.3,
      tissue = data_config$tissue %||% "large intestine",
      species = data_config$species %||% "human",
      max_workers = llm_config$max_workers %||% 3,
      provider = llm_config$provider %||% "openrouter",
      validator_involvement = config$validator$default %||% "v1"
    )
    batch_results_file <- paste0(batch_output, "_full.csv")
  }

  # Run merging tests
  start_time <- Sys.time()
  errors <- list()
  status <- "error"
  merge_results <- list()

  tryCatch({
    # Test 1: Single detail level merge (broad)
    cat("\n--- Test 1: Single detail level merge (broad) ---\n")
    broad_output <- file.path(results_dir, "merge_broad.csv")

    result_broad <- CASSIA::merge_annotations(
      csv_path = batch_results_file,
      output_path = broad_output,
      provider = llm_config$provider %||% "openrouter",
      model = llm_config$model %||% "google/gemini-2.5-flash",
      detail_level = "broad",
      batch_size = 10
    )

    if (!is.null(result_broad) && "Merged_Grouping_1" %in% names(result_broad)) {
      cat("  Broad merge: SUCCESS\n")
      cat("  Output rows:", nrow(result_broad), "\n")
      cat("  Sample groupings:\n")
      cluster_col <- if ("Cluster ID" %in% names(result_broad)) "Cluster ID" else "True Cell Type"
      for (i in 1:min(3, nrow(result_broad))) {
        cat("   ", result_broad[[cluster_col]][i], "->", result_broad$Merged_Grouping_1[i], "\n")
      }
      merge_results$broad <- list(status = "success", output_file = broad_output, num_rows = nrow(result_broad))
    } else {
      errors <- c(errors, "Broad merge failed - missing Merged_Grouping_1 column")
      merge_results$broad <- list(status = "failed")
    }

    # Test 2: Single detail level merge (detailed)
    cat("\n--- Test 2: Single detail level merge (detailed) ---\n")
    detailed_output <- file.path(results_dir, "merge_detailed.csv")

    result_detailed <- CASSIA::merge_annotations(
      csv_path = batch_results_file,
      output_path = detailed_output,
      provider = llm_config$provider %||% "openrouter",
      model = llm_config$model %||% "google/gemini-2.5-flash",
      detail_level = "detailed",
      batch_size = 10
    )

    if (!is.null(result_detailed) && "Merged_Grouping_2" %in% names(result_detailed)) {
      cat("  Detailed merge: SUCCESS\n")
      cat("  Output rows:", nrow(result_detailed), "\n")
      cat("  Sample groupings:\n")
      cluster_col <- if ("Cluster ID" %in% names(result_detailed)) "Cluster ID" else "True Cell Type"
      for (i in 1:min(3, nrow(result_detailed))) {
        cat("   ", result_detailed[[cluster_col]][i], "->", result_detailed$Merged_Grouping_2[i], "\n")
      }
      merge_results$detailed <- list(status = "success", output_file = detailed_output, num_rows = nrow(result_detailed))
    } else {
      errors <- c(errors, "Detailed merge failed - missing Merged_Grouping_2 column")
      merge_results$detailed <- list(status = "failed")
    }

    # Test 3: All detail levels merge (parallel)
    cat("\n--- Test 3: All detail levels merge (parallel) ---\n")
    all_output <- file.path(results_dir, "merge_all.csv")

    result_all <- CASSIA::merge_annotations_all(
      csv_path = batch_results_file,
      output_path = all_output,
      provider = llm_config$provider %||% "openrouter",
      model = llm_config$model %||% "google/gemini-2.5-flash",
      batch_size = 10
    )

    expected_columns <- c("Merged_Grouping_1", "Merged_Grouping_2", "Merged_Grouping_3")
    if (!is.null(result_all) && all(expected_columns %in% names(result_all))) {
      cat("  All levels merge: SUCCESS\n")
      cat("  Output rows:", nrow(result_all), "\n")
      cat("  Columns:", paste(expected_columns, collapse = ", "), "\n")
      cat("\n  Sample comparison:\n")
      cluster_col <- if ("Cluster ID" %in% names(result_all)) "Cluster ID" else "True Cell Type"
      for (i in 1:min(3, nrow(result_all))) {
        cat("   ", result_all[[cluster_col]][i], ":\n")
        cat("      Broad:        ", result_all$Merged_Grouping_1[i], "\n")
        cat("      Detailed:     ", result_all$Merged_Grouping_2[i], "\n")
        cat("      Very Detailed:", result_all$Merged_Grouping_3[i], "\n")
      }
      merge_results$all <- list(status = "success", output_file = all_output, num_rows = nrow(result_all), columns = expected_columns)
    } else {
      missing <- expected_columns[!expected_columns %in% names(result_all)]
      errors <- c(errors, paste("All levels merge failed - missing columns:", paste(missing, collapse = ", ")))
      merge_results$all <- list(status = "failed")
    }

    # Determine overall status
    statuses <- sapply(merge_results, function(r) r$status)
    if (all(statuses == "success")) {
      status <- "passed"
    } else if (any(statuses == "success")) {
      status <- "partial"
    } else {
      status <- "failed"
    }

  }, error = function(e) {
    errors <<- c(errors, e$message)
    status <<- "error"
    cat("\nError:", e$message, "\n")
  })

  duration <- as.numeric(difftime(Sys.time(), start_time, units = "secs"))

  # Save metadata and results
  metadata <- create_test_metadata(
    test_name = "merging_annotation",
    config = config,
    duration_seconds = duration,
    status = status,
    clusters_tested = list("all"),
    errors = errors
  )
  save_test_metadata(results_dir, metadata)

  save_test_results(results_dir, list(
    batch_results_file = batch_results_file,
    merge_results = merge_results,
    detail_levels_tested = c("broad", "detailed", "very_detailed", "all")
  ))

  # Print final result
  success <- status == "passed"
  print_test_result(success, paste("Duration:", round(duration, 2), "s"))

  return(success)
}

# Run test
success <- run_merging_annotation_test()
quit(status = if (success) 0 else 1)
